base/generator.jl

```diff
@@ -0,0 +1,21 @@
+"""
+    Generator(f, iter)
+
+Given a function `f` and an iterator `iter`, construct an iterator that yields
+the values of `f` applied to the elements of `iter`.
+The syntax `f(x) for x in iter` is syntax for constructing an instance of this
+type.
+"""
+immutable Generator{I,F}
```

Comment:
Yes, I think `Generator(func, range)` is a bit more natural since it matches the order of comprehensions, but I suspect dispatching on the kind of iterator will be much more common than dispatching on the type of function. For example below I dispatch on `Generator{IteratorND}`.